### ISSUE TITLE
Test filtering not working correctly in certain cases involving explicit

### ISSUE THREAD

**Issue by CharliePoole** [nunit/nunit#2226]
This issue came up in discussing nunit/nunit-console#134 =
The following example filter expression at the command-line causes all tests to be run, including explicit tests.

```
--where cat==Integration || cat != Integration
```

The expected behavior is that All tests will be executed __except__ Explicit tests that are not in the Integration category. The actual behavior is that all tests are executed.

Possible lines of attack:
* How is the or operator handling explicit selection?
* Are we identifying the `!=` and `!~` operators (and synonyms like `ne` ) correctly as exclusions?

Ideally, whoever does this should add a lot of tests covering explicit selection since this is an area where problems continue to show up.

**Comment by jnm2** [comment]
Oh. Last week makes more sense now. :smile:

**Comment by OmicronPersei** [comment]
Looking to pick this one up.  What part(s) of the code base should I look at to get a feel for where command line args are parsed, etc?  Thanks :)

**Comment by CharliePoole** [comment]
You can look in nunit-console or nunitlite at TestSelectionParser.cs to see how filters are created. But this issue is about what happens in the framework when the filters are applied, not about how they are generated.

The filters are all in the `NUnit.Framework.Internal.Filters` namespace, with tests for each type of filter in the same place in the test assembly. We don't have many (or any?) tests that combine different types of filters in the ways that are causing problems, so you will need to add them to get to the bottom of this issue.

**Comment by OmicronPersei** [comment]
Alright.  I've started so far off with this test, which fails as anticipated:
```
[Test]
public void TestExplicitNotSelected()
{
	//where cat==Dummy || cat!=Dummy
	var filter = new OrFilter(new NotFilter(new CategoryFilter("Dummy")), new CategoryFilter("Dummy"));

	Assert.False(filter.Match(_explicitFixture));
}
```
Just making sure this is what's generally being looked for (along with a lot more tests to verify more combos of filters)?

**Comment by OmicronPersei** [comment]
I'm also seeing the `NotFilter` itself not really working properly in terms of Explicit selection.  The below test fails as well:
```C#
[Test]
public void ExplicitAttributeMatchTest()
{
	var filter = new NotFilter(new CategoryFilter("Dummy"));

	Assert.False(filter.Match(_explicitFixture));
}
```

**Comment by OmicronPersei** [comment]
After reading more, I realized that there's a `CompositeFilter.IsExplicitMatch()` method... oops 😆 .  Updating the two aforementioned test to instead use the `IsExplicitMatch` check, both tests pass.  It would seem that the filters themselves are not the cause of the example bug in the OP, but something else...

**Comment by CharliePoole** [comment]
Primary tests should be of Pass and IsExplicitMatch for these external tests because those are in the interface used by the framework. Match is useful to test for a single filter, because it is the heart of the filter implementation.

**Comment by CharliePoole** [comment]
An explicit test is run off both Pass and Is expliciT match succeed.

**Comment by OmicronPersei** [comment]
So you're saying when testing tests that are marked with `Explicit`, I should be checking whether or not both `Pass()` and `IsExplicitMatch()` succeed?  If so, is there a reason why we don't have our default "Pass" method test both conditions by default?

**Comment by CharliePoole** [comment]
Because the two tests have to be made at two different times. It's a compromise I created to solve the special problem presented by explicit. A test can pass, which means we consider it as a part of the test run, but fail IsExplicitMatch, which causes it to be reported as skipped. Here's a decision table:

Pass? | Explicit Test? | Explicit Match? | Action
------|--------------|-----------------|-------
No    |          -          |           -             | Skip without reporting
Yes   |         No        |           -             | Run the test
Yes   |         Yes       |            No         | Skip and report as explicit
Yes   |         Yes       |           Yes          |  Run the test

**Comment by OmicronPersei** [comment]
Okay makes sense.  Unfortunately it does not seem that the bug lies within the filtering framework bit.  At least according to the fact that the below test passes:
```C#
[Test]
public void TestExplicitNotSelectedCompoundFilter()
{
	//where cat==Dummy || cat!=Dummy
	var filter = new OrFilter(new NotFilter(new CategoryFilter("Dummy")), new CategoryFilter("Dummy"));

	Assert.That(filter.Pass(_explicitFixture));
	Assert.False(filter.IsExplicitMatch(_explicitFixture));
}
```
where `_explicitFixture` is a fixture with the `Explicit` attribute.

Any other ideas where the bug could lurk? 😛

**Comment by CharliePoole** [comment]
Can you still verify the original report at the command-line? Maybe the code has changed since it was reported. If so, then I'd use the debugger to see what is happening.

**Comment by CharliePoole** [comment]
No branch or PR for this one?

**Comment by CharliePoole** [comment]
I ran tests using both nunit3-console and nunitlite-runner. I did it with non-existent and existing categories. I can't reproduce this bug, so I'm closing.

I copied this report from another issue and should have marked it as requiring confirmation. It's always good to make sure that you have a confirmed issue when working on something.

@OmicronPersei If you have new tests that would benefit us, feel free to do a PR anyway.
